DIAMOND implementation

elizabethmcd opened the pull request:
This PR adds support for DIAMOND to run `diamond blastp` of a set of query proteins against a database. The DIAMOND databases  used in the test profiles are the predicted proteins from each metagenomic assembly pooled together and is available in the `Arcadia-Science/test-datasets` repository. The arguments added to the schema is the path to the DIAMOND database (which must be pre-built before using the workflow) and the columns to include in the DIAMOND output. The test profiles do not include columns for taxonomy since the test DIAMOND databases don't have taxonomy information, but that can be added by the user.

elizabethmcd commented:
There might be a potential issue with the long reads implementation of DIAMOND having to do with frameshift penalty scores that I'm playing around with, possibly a test dataset issue as well

elizabethmcd commented:
Update - removed the `--long-reads` and frameshift penalty for DIAMOND blastp of long read assemblies. Previously I've only successfully been able to use this option when doing `diamond blastx` of a DIAMOND database against assembled contigs, not predicted proteins. Which is something we can consider because I'm uncertain with how things are output now if that will extend to contig-level classification with something like MEGAN since we are directly feeding the protein fasta files. But for now, this works for outputting was is needed for HGT exploration

mertcelebi commented on nextflow_schema.json:
```diff
@@ -47,6 +54,16 @@
+                },
+                "diamond_columns": {
+                    "type": "string",
+                    "default": null,
```
should this have sensible defaults?

mertcelebi replied:
if so, i'd also add these defaults to the main config file

taylorreiter replied:
oh yes can we make the defaults either typical `outfmt 6`, or i think better yet since this is serving our purposes, all of the ones we highlighted previously? Or Or Or all possible would be good too, because why not. i don't think it costs  most to just output all the info

taylorreiter replied:
here's all of them if you want to go that route: https://github.com/bbuchfink/diamond/blob/7c86d5e6a357ba666f2939e6955a6cc9abb905f3/src/output/blast_tab_format.cpp#L48

mertcelebi commented on workflows/nanopore.nf:
```diff
@@ -135,6 +142,15 @@ workflow NANOPORE {
+        ch_proteins,
+        ch_diamond_db,
+        "txt",
+        params.diamond_columns
```
I assume this should also use `val_diamond_columns`

elizabethmcd replied:
that's what the params.diamond_columns is referring to, I've made that something the user has to input

elizabethmcd replied:
addressed

mertcelebi reviewed (commented):
It'd be good to include details to the README on the diamond related params etc

taylorreiter commented on conf/test_illumina.config:
```diff
@@ -23,4 +23,6 @@ params {
     platform        = 'illumina'
     sourmash_dbs    = 'https://raw.githubusercontent.com/Arcadia-Science/test-datasets/main/metagenomics/sourmash_dbs/sourmash_dbs_test.csv'
+    diamond_db      = 'https://github.com/Arcadia-Science/test-datasets/raw/main/metagenomics/diamond_dbs/test-cheese-proteins-db.dmnd'
+    diamond_columns = 'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore corrected_bitscore qlen slen'
```
```suggestion
    diamond_columns = 'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore corrected_bitscore qlen slen qcovs qcovhsp'
```

realized two more that would be useful!

taylorreiter commented on conf/test_nanopore.config:
```diff
@@ -23,4 +23,6 @@ params {
     platform        = 'nanopore'
     sourmash_dbs    = 'https://raw.githubusercontent.com/Arcadia-Science/test-datasets/main/metagenomics/sourmash_dbs/sourmash_dbs_test.csv'
+    diamond_db      = 'https://github.com/Arcadia-Science/test-datasets/raw/main/metagenomics/diamond_dbs/test-cheese-proteins-db.dmnd'
+    diamond_columns = 'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore corrected_bitscore qlen slen'
```
```suggestion
    diamond_columns = 'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore corrected_bitscore qlen slen qcovs qcovhsp'
```

elizabethmcd replied:
I tried this out and decided to leave these out because it threw an error with the test data, so the default options and test profile diamond columns are the simplest set of output columns

taylorreiter reviewed (commented):
also commenting -- this looks like a super promising start, just need to iron how dealing with the diamond column headers and specifying a sensible default

elizabethmcd commented:
I have updated by providing default DIAMOND columns in the `nextflow.config` file. For test profiles since this is test data not all of the columns are possible to output, but those can be added for full runs when we are using the uniprot diamond database for example. I also cleaned up how `ch_diamond_columns` is called to be consistent

taylorreiter reviewed (approved):
LGTM!
